Remotion/Core/Development.Analyzers.IntegrationTests/RMCORE0001_SafeContextAnalyzerTests.cs

```diff
@@ -0,0 +1,236 @@
+      var diagnosticAction = Verifier.Diagnostic(Analyzer.AlternativeDescriptor)
+          .WithSpan(11, 5, 11, 33)
+          .WithMessage(
+              "'System.Threading.Tasks.Task.Run' should not be used with SafeContext, use 'Remotion.Context.SafeContext.Task.Run' instead.");
+
+      await (Verifier.VerifyAnalyzerAsync(input, true,diagnosticFunc, diagnosticAction));
+    }
+
+    [Test]
+    public async Task SafeContext_WithTaskContinueWith_Diagnostic ()
+    {
+      const string input = @"
+using System.Threading.Tasks;
+using Remotion.Context;
+public class A
+{
+  public void Test ()
+  {
+    SafeContext.Instance.SetData(""a"", 3);
+
+    var t = Task.CompletedTask;
+    t.ContinueWith((task => { }));
+  }
+}
+";
+      var diagnostic = Verifier.Diagnostic(Analyzer.AlternativeDescriptor)
+          .WithSpan(11, 5, 11, 34)
+          .WithMessage(
+              "'System.Threading.Tasks.Task.ContinueWith' should not be used with SafeContext, use 'Remotion.Context.SafeContext.Task.ContinueWith' instead.");
+
+      await Verifier.VerifyAnalyzerAsync(input,true,diagnostic);
+    }
+
+    [Test]
+    public async Task SafeContext_WithExecutionContextRun_Diagnostic ()
+    {
+      const string input = @"
+using System.Threading;
+using Remotion.Context;
+public class A
+{
+  public void Test ()
+  { 
+    SafeContext.Instance.SetData(""a"", 3);
+    var ec = ExecutionContext.Capture();
+
+    ExecutionContext.Run(
+        ec,
+        _ => { SafeContext.Instance.SetData(""a"", 3); },
+        null);
+  }
+}
+";
+      var diagnosticFunc = Verifier.Diagnostic(Analyzer.AlternativeDescriptor)
+          .WithSpan(11, 5, 14, 14)
+          .WithMessage(
+              "'System.Threading.ExecutionContext.Run' should not be used with SafeContext, use 'Remotion.Context.SafeContext.ExecutionContext.Run' instead.");
+
+      await Verifier.VerifyAnalyzerAsync(input, true,diagnosticFunc);
+    }
+
+    [Test]
+    public async Task SafeContext_WithTasksParallel_Diagnostic ()
+    {
+      const string input = @"
+using System;
+using System.Threading.Tasks;
+using System.Linq;
+using Remotion.Context;
+public class A
+{
+  public void Test ()
+  { 
+    SafeContext.Instance.SetData(""a"", 3);
+
+    Parallel.For(0, 5, i => { i = (int) SafeContext.Instance.GetData(""a"");});
+
+    Parallel.ForEach(Enumerable.Range(5, 10).ToList(), i => Console.Write(i));
+
+    Parallel.ForEachAsync(
+        Enumerable.Range(0, 5),
+        new ParallelOptions(), 
+        (i, token) => ValueTask.CompletedTask
+    );
+
+    Parallel.Invoke(() => Console.Write(""a""));
+  }
+}
+";
+      var diagnosticFor = Verifier.Diagnostic(Analyzer.AlternativeDescriptor)
+          .WithSpan(12, 5, 12, 77)
+          .WithMessage(
+              "'System.Threading.Tasks.Parallel.For' should not be used with SafeContext by itself, use 'provider.OpenSafeContextBoundary()' to create a safe context boundary.");
```

Comment:
We have an API for parallel called `SafeContext.Parallel.OpenSafeContextBoundary` that should be used instead of `provider.OpenSafeContextBoundary` so we suggest that here